python/simulation.py

```diff
@@ -859,6 +864,184 @@ def amplitude(self, point, component):
+         (may be 0d, 1d, 2d, or 3d). No default.
+        + **`center` [`Vector3` class]** — Alternative method for specifying volume, using a center point
+        + **`size` [`Vector3` class]** — Alternative method for specifying volume, using a size vector
+        + **`m` [`Optional[int]`]** — Directly pecifies the order of the numerator $P$. If not specified,
+         defaults to the length of aggregated field data times `m_frac`.
+        + **`n` [`Optional[int]`]** — Specifies the order of the denominator $Q$. Defaults
+         to length of field data - m - 1.
+        + **`m_frac` [`float`]** — Method for specifying `m` as a fraction of
+         field samples to use as order for numerator. Default is 0.5.
+        + **`n_frac` [`Optional[float]`]** — Fraction of field samples to use as order for
+         denominator. No default.
+        + **`sampling_interval` [`int`]** — Specifies the interval at which to sample the field data.
+         Defaults to 1.
+        + **`start_time` [`int`]** — Specifies the time (in increments of dt) at which
+         to start sampling the field data. Default 0 (beginning of simulation).
+        + **`stop_time` [`Optional[int]`]** — Specifies the time (in increments of dt) at which
+         to stop sampling the field data. Default is `None` (end of simulation).
+        """
+        self.c = c
+        self.vol = vol
+        self.center = center
+        self.size = size
+        self.m = m
+        self.n = n
+        self.m_frac = m_frac
+        self.n_frac = n_frac
+        self.sampling_interval = sampling_interval
+        self.start_time = start_time
+        self.stop_time = stop_time
+        self.data = []
+        self.data_dt = 0
+        self.dft: Callable = None
+        self.step_func = self._pade()
+
+    def __call__(self, sim, todo):
+        """
+        Allows a Pade instance to be used as a step function.
+        """
+        self.step_func(sim, todo)
+
+    def _collect_pade(self):
+        def _collect1(c, vol, center, size):
+            self.t0 = 0
+
+            def _collect2(sim):
+                self.data_dt = sim.meep_time() - self.t0
+                self.t0 = sim.meep_time()
+                self.data.append(sim.get_array(c, vol, center, size))
+
+            return _collect2
+
+        return _collect1
+
+    def _analyze_pade(self, sim):
+        # Ensure that the field data has dimension (# time steps x (volume dims))
+        self.data = np.atleast_2d(self.data)
+
+        # If the supplied volume is actually a point, np.atleast_2d will have
+        # shape (1, T), we want (T, 1)
+        if np.shape(self.data)[0] == 1:
+            self.data = self.data.T
+
+        # Sample the collected field data and possibly truncate the beginning or
+        # end to avoid transients
+        samples = np.array(
+            self.data[self.start_time : self.stop_time : self.sampling_interval]
+        )
```

Comment:
Let's add a TODO for enhancing performance by rewriting this section using a custom run function to collect only the necessary field points (we have all the machinery) rather than collecting the fields at every timestep.

Reply:
For formatting, see: https://google.github.io/styleguide/pyguide.html#s3.12-todo-comments